File: src/protocol/j1939.c
#include "j1939.h"
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <stdio.h>

// J1939 ID structure: PPP RRRR PPPPPPPP PPPPPPPP SSSSSSSS
// P = Priority (3 bits), R = Reserved/DP (1 bit), PGN (18 bits), S = Source (8 bits)

uint32_t j1939_get_pgn(uint32_t can_id) {
    // PDU Format (PF) is bits 16-23
    uint8_t pf = (can_id >> 16) & 0xFF;

    if (pf < 240) {
        // PDU1 format: PGN = PF * 256 (PS is destination)
        return ((uint32_t)pf) << 8;
    } else {
        // PDU2 format: PGN = PF * 256 + PS
        uint8_t ps = (can_id >> 8) & 0xFF;
        return (((uint32_t)pf) << 8) | ps;
    }
}

uint8_t j1939_get_source(uint32_t can_id) {
    return can_id & 0xFF;
}

uint32_t j1939_build_id(uint32_t pgn, uint8_t priority, uint8_t source) {
    // Priority in bits 26-28, PGN in bits 8-25, source in bits 0-7
    return ((uint32_t)(priority & 0x07) << 26) |
           ((pgn & 0x3FFFF) << 8) |
           source;
}

// Convert J1939 temperature byte to Celsius (1°C/bit, +40 offset)
static float decode_temp_byte(uint8_t byte) {
    if (byte == 0xFF) return -999.0f;  // Not available
    return (float)byte - 40.0f;
}

// Convert J1939 16-bit temperature to Celsius (0.03125 deg/bit, -273 offset)
// J1939 uses little-endian (low byte first)
static float decode_temp_16bit_le(uint8_t low, uint8_t high) {
    if (high == 0xFF && low == 0xFF) return -999.0f;
    uint16_t val = ((uint16_t)high << 8) | low;
    return (float)val * 0.03125f - 273.0f;
}

// Convert J1939 pressure byte to kPa (2 kPa/bit)
static float decode_pressure_2kpa(uint8_t byte) {
    if (byte == 0xFF) return -999.0f;
    return (float)byte * 2.0f;
}

// Convert J1939 pressure byte to kPa (4 kPa/bit)
static float decode_pressure_4kpa(uint8_t byte) {
    if (byte == 0xFF) return -999.0f;
    return (float)byte * 4.0f;
}

// Convert J1939 barometric pressure to kPa (0.5 kPa/bit)
static float decode_baro_pressure(uint8_t byte) {
    if (byte == 0xFF) return -999.0f;
    return (float)byte * 0.5f;
}

// Convert humidity (0.4%/bit)
static float decode_humidity(uint8_t byte) {
    if (byte == 0xFF) return -999.0f;
    return (float)byte * 0.4f;
}

static uint32_t get_time_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint32_t)(ts.tv_sec * 1000 + ts.tv_nsec / 1000000);
}

bool j1939_parse_message(const TCanFrame* frame, TSensorData* data) {
    if (!frame->extended) return false;

    uint8_t source = j1939_get_source(frame->id);
    if (source != OSSM_SOURCE_ADDRESS) return false;

    uint32_t pgn = j1939_get_pgn(frame->id);
    const uint8_t* d = frame->data;
    bool updated = false;

    switch (pgn) {
        case PGN_AMBIENT_CONDITIONS:  // 65269
            // OSSM encoding (J1939 little-endian):
            // buf[0] = barometric * 2 (0.5 kPa/bit)
            // buf[3-4] = ambient temp 16-bit LE (0.03125°C/bit, +273 offset)
            // buf[5] = air inlet temp (1°C/bit, +40 offset)
            data->absoluteBarometricPressurekPa = decode_baro_pressure(d[0]);
            data->ambientTemperatureC = decode_temp_16bit_le(d[3], d[4]);
            data->airInletTemperatureC = decode_temp_byte(d[5]);
            updated = true;
            break;

        case PGN_INLET_EXHAUST:  // 65270
            // OSSM encoding (J1939 little-endian):
            // buf[1] = boost pressure / 2 (2 kPa/bit)
            // buf[2] = air inlet temp + 40 (boost temp, 1°C/bit)
            // buf[3] = air inlet pressure / 2 (2 kPa/bit)
            // buf[5-6] = EGT 16-bit LE (0.03125°C/bit, +273 offset)
            data->boostPressurekPa = decode_pressure_2kpa(d[1]);
            data->boostTemperatureC = decode_temp_byte(d[2]);
            data->airInletPressurekPa = decode_pressure_2kpa(d[3]);
            data->egtTemperatureC = decode_temp_16bit_le(d[5], d[6]);
            updated = true;
            break;

        case PGN_ENGINE_TEMP:  // 65262
            // OSSM encoding (J1939 little-endian):
            // buf[0] = coolant temp + 40 (1°C/bit)
            // buf[1] = fuel temp + 40 (1°C/bit)
            // buf[2-3] = oil temp 16-bit LE (0.03125°C/bit, +273 offset)
            data->coolantTemperatureC = decode_temp_byte(d[0]);
            data->fuelTemperatureC = decode_temp_byte(d[1]);
            data->oilTemperatureC = decode_temp_16bit_le(d[2], d[3]);
            updated = true;
            break;

        case PGN_ENGINE_FLUID_PRESS:  // 65263
            // OSSM encoding:
            // buf[0] = fuel pressure / 4 (4 kPa/bit)
            // buf[3] = oil pressure / 4 (4 kPa/bit)
            // buf[6] = coolant pressure / 2 (2 kPa/bit)
            data->fuelPressurekPa = decode_pressure_4kpa(d[0]);
            data->oilPressurekPa = decode_pressure_4kpa(d[3]);
            data->coolantPressurekPa = decode_pressure_2kpa(d[6]);
            updated = true;
            break;

        case PGN_ENGINE_TEMP_2:  // 65129
            // OSSM encoding:
            // buf[0-1] = intake manifold 1 temp 16-bit big-endian (hi-res boost temp)
            // buf[2-3] = coolant temp 16-bit big-endian (hi-res)
            // Note: we already get boost temp from 65270, this is hi-res version
            // data->boostTemperatureC = decode_temp_16bit_be(d[0], d[1]);
            updated = true;
            break;

        case PGN_ENGINE_TEMP_3:  // 65189
            // OSSM encoding:
            // buf[0] = intake manifold 2 temp (CAC inlet, 1°C/bit, +40)
            // buf[1] = intake manifold 3 temp (transfer pipe, 1°C/bit, +40)
            // buf[2] = intake manifold 4 temp (air inlet, 1°C/bit, +40)
            data->cacInletTemperatureC = decode_temp_byte(d[0]);
            data->transferPipeTemperatureC = decode_temp_byte(d[1]);
            updated = true;
            break;

        case PGN_TURBO_PRESS:  // 65190
            // OSSM encoding (J1939 little-endian):
            // buf[0-1] = turbo 1 boost 16-bit LE (0.125 kPa/bit)
            // buf[2-3] = turbo 2 boost 16-bit LE (0.125 kPa/bit)
            if (d[0] != 0xFF || d[1] != 0xFF) {
                uint16_t val = ((uint16_t)d[1] << 8) | d[0];
                data->cacInletPressurekPa = (float)val * 0.125f;
            }
            if (d[2] != 0xFF || d[3] != 0xFF) {
                uint16_t val = ((uint16_t)d[3] << 8) | d[2];
                data->transferPipePressurekPa = (float)val * 0.125f;
            }
            updated = true;
            break;

        case 65164:  // PGN_SUPPLY_PRESSURE - OSSM uses for humidity & engine bay
            // OSSM encoding:
            // buf[0] = engine bay temp + 40 (1°C/bit)
            // buf[6] = humidity / 0.4 (0.4%/bit)
            data->engineBayTemperatureC = decode_temp_byte(d[0]);
            data->humidity = decode_humidity(d[6]);
            updated = true;
            break;

        default:
            break;
    }

    if (updated) {
        data->lastUpdateMs = get_time_ms();
    }

    return updated;
}

int j1939_send_command(int sock, EOssmCommand cmd, const uint8_t* params, uint8_t param_len) {
    TCanFrame frame;

    frame.id = j1939_build_id(PGN_OSSM_COMMAND, 6, 0x00);  // Priority 6, source 0
    frame.extended = true;
    frame.len = 8;

    memset(frame.data, 0xFF, 8);
    frame.data[0] = (uint8_t)cmd;

    if (params && param_len > 0) {
        if (param_len > 7) param_len = 7;
        memcpy(&frame.data[1], params, param_len);
    }

    return socketcan_send(sock, &frame);
}

int j1939_check_response(int sock, uint8_t expected_cmd, uint8_t* response_data, uint8_t* response_len) {
    TCanFrame frame;
    int timeout_ms = 1000;
    int elapsed = 0;

    // Small delay to let OSSM process the command
    usleep(50000);  // 50ms

    while (elapsed < timeout_ms) {
        int result = socketcan_receive(sock, &frame);
        if (result > 0) {
            uint32_t pgn = j1939_get_pgn(frame.id);
            uint8_t source = j1939_get_source(frame.id);

            // Debug: uncomment to see what we're receiving
            // fprintf(stderr, "RX: ID=%08X PGN=%u SRC=%02X CMD=%02X\n",
            //         frame.id, pgn, source, frame.data[0]);

            if (pgn == PGN_OSSM_RESPONSE && source == OSSM_SOURCE_ADDRESS) {
                if (frame.data[0] == expected_cmd) {
                    uint8_t error = frame.data[1];
                    if (response_data && response_len) {
                        *response_len = 6;
                        memcpy(response_data, &frame.data[2], 6);
                    }
                    return error;
                }
            }
        }
        usleep(5000);  // 5ms
        elapsed += 5;
    }

    return -1;  // Timeout
}

int j1939_enable_spn(int sock, uint16_t spn, bool enable, uint8_t input) {
    // OSSM format: [cmd, spn_high, spn_low, enable, input]
    uint8_t params[4];
    params[0] = (spn >> 8) & 0xFF;  // spn_high
    params[1] = spn & 0xFF;          // spn_low
    params[2] = enable ? 1 : 0;      // enable
    params[3] = input;               // input

    #ifdef DEBUG
    fprintf(stderr, "TX enable_spn: SPN=%u enable=%d input=%d -> [%02X %02X %02X %02X]\n",
            spn, enable, input, params[0], params[1], params[2], params[3]);
    #endif

    if (j1939_send_command(sock, CMD_ENABLE_SPN, params, 4) < 0) {
        return -1;
    }

    return j1939_check_response(sock, CMD_ENABLE_SPN, NULL, NULL);
}

int j1939_set_ntc_preset(int sock, uint8_t input, ENtcPreset preset) {
    uint8_t params[2];
    params[0] = input;
    params[1] = (uint8_t)preset;

    if (j1939_send_command(sock, CMD_NTC_PRESET, params, 2) < 0) {
        return -1;
    }

    return j1939_check_response(sock, CMD_NTC_PRESET, NULL, NULL);
}

int j1939_set_pressure_preset(int sock, uint8_t input, EPressurePreset preset) {
    uint8_t params[2];
    params[0] = input;
    params[1] = (uint8_t)preset;

    if (j1939_send_command(sock, CMD_PRESSURE_PRESET, params, 2) < 0) {
        return -1;
    }

    return j1939_check_response(sock, CMD_PRESSURE_PRESET, NULL, NULL);
}

int j1939_set_tc_type(int sock, uint8_t type) {
    uint8_t params[1];
    params[0] = type;

    if (j1939_send_command(sock, CMD_SET_TC_TYPE, params, 1) < 0) {
        return -1;
    }

    return j1939_check_response(sock, CMD_SET_TC_TYPE, NULL, NULL);
}

int j1939_query_config(int sock, TConfigState* state) {
    uint8_t params[1] = {0};  // Query type 0 = SPN counts
    uint8_t response[6];
    uint8_t len;

    if (j1939_send_command(sock, CMD_QUERY, params, 1) < 0) {
        return -1;
    }

    int result = j1939_check_response(sock, CMD_QUERY, response, &len);
    if (result == 0 && state) {
        state->tempCount = response[0];
        state->pressureCount = response[1];
        state->egtEnabled = response[2] != 0;
        state->bme280Enabled = response[3] != 0;
    }

    return result;
}

int j1939_query_spn_assignments(int sock, TSpnAssignments* assignments) {
    uint8_t response[6];
    uint8_t len;
    int result;

    // Initialize to zero
    memset(assignments, 0, sizeof(TSpnAssignments));

    // Query temp SPNs (3 sub-queries for 8 inputs)
    for (uint8_t subQuery = 0; subQuery < 3; subQuery++) {
        uint8_t params[2] = {1, subQuery};  // Query type 1 = temp SPNs
        if (j1939_send_command(sock, CMD_QUERY, params, 2) < 0) {
            return -1;
        }
        result = j1939_check_response(sock, CMD_QUERY, response, &len);
        if (result != 0) return result;

        // Parse 3 SPNs from response (high byte, low byte pairs)
        for (uint8_t i = 0; i < 3; i++) {
            uint8_t idx = subQuery * 3 + i;
            if (idx < 8) {
                uint16_t spn = ((uint16_t)response[i * 2] << 8) | response[i * 2 + 1];
                if (spn != 0xFFFF) {
                    assignments->tempSpns[idx] = spn;
                }
            }
        }
    }

    // Query pressure SPNs (3 sub-queries for 7 inputs)
    for (uint8_t subQuery = 0; subQuery < 3; subQuery++) {
        uint8_t params[2] = {2, subQuery};  // Query type 2 = pressure SPNs
        if (j1939_send_command(sock, CMD_QUERY, params, 2) < 0) {
            return -1;
        }
        result = j1939_check_response(sock, CMD_QUERY, response, &len);
        if (result != 0) return result;

        // Parse 3 SPNs from response
        for (uint8_t i = 0; i < 3; i++) {
            uint8_t idx = subQuery * 3 + i;
            if (idx < 7) {
                uint16_t spn = ((uint16_t)response[i * 2] << 8) | response[i * 2 + 1];
                if (spn != 0xFFFF) {
                    assignments->presSpns[idx] = spn;
                }
            }
        }
    }

    return 0;
}

int j1939_save_config(int sock) {
    if (j1939_send_command(sock, CMD_SAVE, NULL, 0) < 0) {
        return -1;
    }

    return j1939_check_response(sock, CMD_SAVE, NULL, NULL);
}

int j1939_reset_config(int sock) {
    if (j1939_send_command(sock, CMD_RESET, NULL, 0) < 0) {
        return -1;
    }

    return j1939_check_response(sock, CMD_RESET, NULL, NULL);
}

File: src/protocol/j1939.ts
// J1939 protocol encoding/decoding for OSSM communication
import { CanBus, CanFrame } from '../can/socketcan';

// OSSM proprietary PGNs
const PGN_COMMAND = 65280;   // 0xFF00 - Commands TO OSSM
const PGN_RESPONSE = 65281;  // 0xFF01 - Responses FROM OSSM

// Standard J1939 PGNs for sensor data
export const PGN = {
  ENGINE_TEMP_1: 65262,      // 0xFEEE - Coolant, fuel, oil temps
  ENGINE_FLUID_PRESS: 65263, // 0xFEEF - Fuel, oil, coolant pressures
  AMBIENT_COND: 65269,       // 0xFEF5 - Baro, ambient temp
  INLET_EXHAUST: 65270,      // 0xFEF6 - Air inlet, EGT, boost
  ENGINE_TEMP_2: 65129,      // 0xFE69 - Additional temps
  TURBO_INFO_1: 65189,       // 0xFEA5 - Turbo temps
  TURBO_INFO_2: 65190,       // 0xFEA6 - Turbo pressures
  EEC6: 65164,               // 0xFE8C - Engine bay temp, humidity
};

// Command IDs
export const CMD = {
  ENABLE_SPN: 1,
  SET_PRESSURE_RANGE: 3,
  SET_TC_TYPE: 4,
  QUERY: 5,
  SAVE: 6,
  RESET: 7,
  NTC_PRESET: 8,
  PRESSURE_PRESET: 9,
};

// Default OSSM source address
const OSSM_SOURCE_ADDRESS = 149;  // 0x95

export interface SensorData {
  // Temperatures (Celsius)
  coolantTemp?: number;
  fuelTemp?: number;
  oilTemp?: number;
  ambientTemp?: number;
  airInletTemp?: number;
  egtTemp?: number;
  boostTemp?: number;
  cacInletTemp?: number;
  transferPipeTemp?: number;
  engineBayTemp?: number;

  // Pressures (kPa)
  oilPressure?: number;
  fuelPressure?: number;
  coolantPressure?: number;
  boostPressure?: number;
  airInletPressure?: number;
  cacInletPressure?: number;

  // Ambient
  barometricPressure?: number;
  humidity?: number;
}

export class J1939Protocol {
  private can: CanBus;
  private responseResolve: ((data: Buffer) => void) | null = null;
  private sensorData: SensorData = {};
  private sensorHandler: ((data: SensorData) => void) | null = null;

  constructor(can: CanBus) {
    this.can = can;
    this.can.onMessage(this.handleFrame.bind(this));
  }

  private handleFrame(frame: CanFrame): void {
    if (!frame.ext) return;  // J1939 uses extended IDs

    const pgn = this.extractPgn(frame.id);
    const sourceAddr = frame.id & 0xFF;

    // Only process frames from OSSM
    if (sourceAddr !== OSSM_SOURCE_ADDRESS) return;

    // Handle command response
    if (pgn === PGN_RESPONSE && this.responseResolve) {
      this.responseResolve(frame.data);
      this.responseResolve = null;
      return;
    }

    // Handle sensor data PGNs
    this.decodeSensorData(pgn, frame.data);
  }

  private extractPgn(canId: number): number {
    // J1939 PGN is in bits 8-25 of the 29-bit ID
    const pf = (canId >> 16) & 0xFF;  // PDU Format
    const ps = (canId >> 8) & 0xFF;   // PDU Specific

    if (pf >= 240) {
      // PDU2 format: PGN = PF * 256 + PS
      return (pf << 8) | ps;
    } else {
      // PDU1 format: PGN = PF * 256
      return pf << 8;
    }
  }

  private buildCanId(pgn: number, sourceAddr: number = 0xFE, priority: number = 6): number {
    // Build 29-bit J1939 CAN ID
    return (priority << 26) | (pgn << 8) | sourceAddr;
  }

  private decodeSensorData(pgn: number, data: Buffer): void {
    switch (pgn) {
      case PGN.ENGINE_TEMP_1:
        // Byte 0: Coolant temp (offset -40)
        // Byte 2: Fuel temp (offset -40)
        // Byte 3: Oil temp (offset -40)
        if (data[0] !== 0xFF) this.sensorData.coolantTemp = data[0] - 40;
        if (data[2] !== 0xFF) this.sensorData.fuelTemp = data[2] - 40;
        if (data[3] !== 0xFF) this.sensorData.oilTemp = data[3] - 40;
        break;

      case PGN.ENGINE_FLUID_PRESS:
        // Byte 0-1: Fuel delivery pressure (4 kPa/bit)
        // Byte 3: Oil pressure (4 kPa/bit)
        // Byte 4: Coolant pressure (2 kPa/bit)
        if (data[3] !== 0xFF) this.sensorData.oilPressure = data[3] * 4;
        if (data[4] !== 0xFF) this.sensorData.coolantPressure = data[4] * 2;
        const fuelPress = data[0] | (data[1] << 8);
        if (fuelPress !== 0xFFFF) this.sensorData.fuelPressure = fuelPress * 4;
        break;

      case PGN.AMBIENT_COND:
        // Byte 0: Barometric pressure (0.5 kPa/bit)
        // Byte 3-4: Ambient temp (0.03125 C/bit, offset -273)
        if (data[0] !== 0xFF) this.sensorData.barometricPressure = data[0] * 0.5;
        const ambientRaw = data[3] | (data[4] << 8);
        if (ambientRaw !== 0xFFFF) this.sensorData.ambientTemp = ambientRaw * 0.03125 - 273;
        break;

      case PGN.INLET_EXHAUST:
        // Byte 1: Boost pressure (2 kPa/bit)
        // Byte 2-3: EGT (0.03125 C/bit, offset -273)
        // Byte 4: Air inlet temp (offset -40)
        // Byte 5: Air inlet pressure (2 kPa/bit)
        if (data[1] !== 0xFF) this.sensorData.boostPressure = data[1] * 2;
        if (data[4] !== 0xFF) this.sensorData.airInletTemp = data[4] - 40;
        if (data[5] !== 0xFF) this.sensorData.airInletPressure = data[5] * 2;
        const egtRaw = data[2] | (data[3] << 8);
        if (egtRaw !== 0xFFFF) this.sensorData.egtTemp = egtRaw * 0.03125 - 273;
        break;

      case PGN.ENGINE_TEMP_2:
        // Byte 0-1: Boost temp (0.03125 C/bit, offset -273)
        const boostTempRaw = data[0] | (data[1] << 8);
        if (boostTempRaw !== 0xFFFF) this.sensorData.boostTemp = boostTempRaw * 0.03125 - 273;
        break;

      case PGN.TURBO_INFO_1:
        // Byte 0: CAC inlet temp (offset -40)
        // Byte 1: Transfer pipe temp (offset -40)
        // Byte 2: Engine bay temp (offset -40)
        if (data[0] !== 0xFF) this.sensorData.cacInletTemp = data[0] - 40;
        if (data[1] !== 0xFF) this.sensorData.transferPipeTemp = data[1] - 40;
        if (data[2] !== 0xFF) this.sensorData.engineBayTemp = data[2] - 40;
        break;

      case PGN.TURBO_INFO_2:
        // Byte 2: CAC inlet pressure (2 kPa/bit)
        if (data[2] !== 0xFF) this.sensorData.cacInletPressure = data[2] * 2;
        break;

      case PGN.EEC6:
        // Byte 0: Engine bay temp (offset -40)
        // Byte 6: Humidity (0.5%/bit)
        if (data[0] !== 0xFF) this.sensorData.engineBayTemp = data[0] - 40;
        if (data[6] !== 0xFF) this.sensorData.humidity = data[6] * 0.5;
        break;
    }

    // Notify listener
    if (this.sensorHandler) {
      this.sensorHandler(this.sensorData);
    }
  }

  onSensorData(handler: (data: SensorData) => void): void {
    this.sensorHandler = handler;
  }

  getSensorData(): SensorData {
    return { ...this.sensorData };
  }

  private async sendCommand(cmdId: number, data: number[] = []): Promise<Buffer> {
    const buf = Buffer.alloc(8, 0xFF);
    buf[0] = cmdId;
    data.forEach((byte, i) => {
      if (i < 7) buf[i + 1] = byte;
    });

    const canId = this.buildCanId(PGN_COMMAND);

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.responseResolve = null;
        reject(new Error('No response from OSSM - check connection'));
      }, 2000);

      this.responseResolve = (responseData) => {
        clearTimeout(timeout);
        resolve(responseData);
      };

      this.can.send({ id: canId, data: buf, ext: true });
    });
  }

  async enableSpn(spn: number, enable: boolean, input: number = 0): Promise<boolean> {
    const spnHi = (spn >> 8) & 0xFF;
    const spnLo = spn & 0xFF;
    const response = await this.sendCommand(CMD.ENABLE_SPN, [spnHi, spnLo, enable ? 1 : 0, input]);
    return response[0] === 0;  // 0 = OK
  }

  async setNtcPreset(input: number, preset: number): Promise<boolean> {
    const response = await this.sendCommand(CMD.NTC_PRESET, [input, preset]);
    return response[0] === 0;
  }

  async setPressurePreset(input: number, preset: number): Promise<boolean> {
    const response = await this.sendCommand(CMD.PRESSURE_PRESET, [input, preset]);
    return response[0] === 0;
  }

  async setThermocoupleType(tcType: number): Promise<boolean> {
    const response = await this.sendCommand(CMD.SET_TC_TYPE, [tcType]);
    return response[0] === 0;
  }

  async query(): Promise<Buffer> {
    return this.sendCommand(CMD.QUERY);
  }

  async save(): Promise<boolean> {
    const response = await this.sendCommand(CMD.SAVE);
    return response[0] === 0;
  }

  async reset(): Promise<boolean> {
    const response = await this.sendCommand(CMD.RESET);
    return response[0] === 0;
  }
}
